### PR TITLE
Add the exists and setAttribute client-worker RPCs

### PR DESCRIPTION
**Pull request by huanghua78**
### What changes are proposed in this pull request?

Implements the exists and setAttribute client-worker RPCs

### Why are the changes needed?

Client completes all operations by RPCs to workers.
Client only does fallback in case of failures when fallback is enabled.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

**Review by yyongycy — APPROVED** [review]
lgtm, it would support metadata op in worker,

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 1]
```diff
@@ -807,6 +809,40 @@ public void createDirectory(String path, CreateDirectoryPOptions options)
...
+      }
+      if (options.hasOwner() && options.hasGroup()) {
+        mUfs.setOwner(path, options.getOwner(), options.getGroup());
+      } else if (options.hasOwner()) {
+        mUfs.setOwner(path, options.getOwner(), null);
+      } else if (options.hasGroup()) {
+        mUfs.setOwner(path, null, options.getOwner());
+      }
+      // Ignore options.hasRecursive()
+      if (options.hasPinned() || options.hasPersisted()
```
Comment (elega) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:835]: can you add a todo to invalidate cache?

Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:835]: OK.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -807,6 +809,40 @@ public void createDirectory(String path, CreateDirectoryPOptions options)
     }
   }
 
+  @Override
+  public boolean exists(String path, ExistsPOptions options) {
+    try {
+      return mUfs.exists(path);
```
[the remark this answers is not in the document]
Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:834]: I added code to use local metadata cache.

Also I added a TODO to you, @elega .
Please help to update local cached metadata if necessary.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 3]
```diff
@@ -807,6 +809,59 @@ public void createDirectory(String path, CreateDirectoryPOptions options)
...
+    // TODO(Yimin) Load/reload metadata from UFS
+    try {
+      return mUfs.exists(path);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void setAttribute(String path, SetAttributePOptions options) {
```
Comment (dbw9580) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:841]: there's a `recursive` option for `setAttribute` to set the same set of attributes to all files under a directory. I don't think we support doing that under Dora, so either we remove that option, or leave a todo to remove that option.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 4]
```diff
@@ -807,6 +809,59 @@ public void createDirectory(String path, CreateDirectoryPOptions options)
...
+      if (options.hasMode()) {
+        mUfs.setMode(path, ModeUtils.protoToShort(options.getMode()));
+      }
+      if (options.hasOwner() && options.hasGroup()) {
+        mUfs.setOwner(path, options.getOwner(), options.getGroup());
+      } else if (options.hasOwner()) {
+        mUfs.setOwner(path, options.getOwner(), null);
+      } else if (options.hasGroup()) {
+        mUfs.setOwner(path, null, options.getOwner());
+      }
```
Comment (dbw9580) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:852]: invalidate metadata cache as the attributes have changed.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Can you please fill me in what was the motivation to support `setAttribute`? If there are some integration tests that test `chmod` and `chown`, we should delete those tests as I don't think Dora wants to be an all-round file system

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 4 above)
Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:852]: Yes. I added TODO here, and  @elega will do that.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 3 above)
Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:841]: We don't support recursive at this moment.
I just removed the checking of this option here.
Maybe it's better to add comments about this.

Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:845]: Added comments.

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:834]: @elega please take another look.

**Comment by huanghua78** [comment]
@dbw9580 please review again.

**Review by dbw9580 — APPROVED** [review]

**Comment by huanghua78** [comment]
Thanks to @dbw9580 and @yyongycy for reviewing this PR.

**Comment by huanghua78** [comment]
alluxio-bot, merge this please

**Pull request merged**